### PR TITLE
Address a few comments by Moritz on statement lifecycle logging

### PR DESCRIPTION
**Pull request by umanwizard**
See [here](https://github.com/MaterializeInc/materialize/pull/24117#pullrequestreview-1878641233) for context.

### CODE REVIEW

**Review by antiguru — APPROVED** [review]
Looks good to me, but I'd like to get @teskje's take on this before merging. Thank you!

File: src/compute-client/src/controller/instance.rs [hunk 1]
```diff
@@ -1532,27 +1532,7 @@ where
...
-                    .map(|state| state.write_frontier.clone());
-
-                if let (Some(old), Some(new)) = (old_upper, new_upper) {
-                    (old != new).then_some(ComputeControllerResponse::FrontierUpper { id, upper })
-                } else {
-                    // this is surprising, but we should already log something in `handle_frontier_upper`,
-                    // so no need to do so here
-                    None
-                }
+                self.handle_frontier_upper(id, upper.clone(), replica_id)
```
Comment (teskje) [inline src/compute-client/src/controller/instance.rs:1535]:
```suggestion
                self.handle_frontier_upper(id, upper, replica_id)
```

File: src/compute-client/src/controller/instance.rs [hunk 2]
```diff
@@ -1570,12 +1550,20 @@ where
...
         id: GlobalId,
         new_frontier: Antichain<T>,
         replica_id: ReplicaId,
-    ) {
+    ) -> Option<ComputeControllerResponse<T>> {
+        let old_upper = self
+            .compute
+            .collection(id)
+            .ok()
+            .map(|state| state.write_frontier.clone());
```
Comment (teskje) [inline src/compute-client/src/controller/instance.rs:1565]: We are getting the collection in the next `let-else`, so if we move this block down we don't have to handle the absence case.

File: src/controller/src/lib.rs [hunk 3]
```diff
@@ -167,8 +168,23 @@ pub struct Controller<T = mz_repr::Timestamp> {
     /// Arguments for secrets readers.
     secrets_args: SecretsReaderCliArgs,
 
-    watch_sets: BTreeMap<GlobalId, Vec<Rc<(T, Box<dyn Any>)>>>,
+    /// A map associating a global ID to a vector of all the watch sets
+    /// that that ID is part of which have not yet been fulfilled for that object.
+    ///
+    /// See [`self.install_watch_se)`] for a description of watch sets.
```
Comment (teskje) [inline src/controller/src/lib.rs:174]:
```suggestion
    /// See [`self.install_watch_set`] for a description of watch sets.
```

File: src/controller/src/lib.rs [hunk 4]
```diff
@@ -167,8 +168,23 @@ pub struct Controller<T = mz_repr::Timestamp> {
...
+    // vector here. That way, when the entire watch set is fulfilled,
+    // the corresponding `Rc` will be the last reference to it, and
+    // the call to `Rc::into_inner` will succeed.
+    objects_to_unfulfilled_watch_sets: BTreeMap<GlobalId, Vec<Rc<(T, Box<dyn Any>)>>>,
 
+    /// A list of watch sets that were already fulfilled as soon as
+    /// they were installed, and thus that must be returned to the
+    /// client on the next call to [`self.process`].
+    ///
+    /// See [`self.install_watch_se)`] for a description of watch sets.
```
Comment (teskje) [inline src/controller/src/lib.rs:187]:
```suggestion
    /// See [`self.install_watch_set`] for a description of watch sets.
```

File: src/controller/src/lib.rs [hunk 5]
```diff
@@ -286,14 +324,55 @@ where
...
                     .or_default()
                     .push(Rc::clone(&state));
             }
         }
     }
 
+    /// Process a pending response from the storage controller. If necessary,
+    /// return a higher-level response to our client.
+    ///
+    /// Precondition: The last readiness status generated must be `Readiness::Storage`.
```
Comment (teskje) [inline src/controller/src/lib.rs:338]: Is this really a necessary precondition? I don't think anything goes wrong if it's not fulfilled, the storage controller just won't have any work to do.

File: src/controller/src/lib.rs [hunk 6]
```diff
@@ -286,14 +324,55 @@ where
...
+            |mz_storage_client::controller::Response::FrontierUpdates(r)| {
+                self.handle_frontier_updates(&r)
+            },
+        ))
+    }
+
+    /// Process a pending response from the compute controller. If necessary,
+    /// return a higher-level response to our client.
+    ///
+    /// Precondition: The last readiness status generated must be `Readiness::Compute`.
```
Comment (teskje) [inline src/controller/src/lib.rs:353]: Same here: I don't think this is a necessary precondition.

**Review by teskje — APPROVED** [review]
Looks good, thanks! I have mostly nits and a documented precondition that I think we don't need.

@umanwizard Lmk if you want me to get this merged!

**Comment by umanwizard** [comment]
TFTR. I no longer have the ability to merge things, so it’s best if you or Moritz take it over.On Mar 4, 2024, at 05:34, Jan Teske ***@***.***> wrote:﻿
@teskje approved this pull request.

Looks good, thanks! I have mostly nits and a documented precondition that I think we don't need.
@umanwizard Lmk if you want me to get this merged!

In src/compute-client/src/controller/instance.rs:
> -                self.handle_frontier_upper(id, upper.clone(), replica_id);
-
-                let new_upper = self
-                    .compute
-                    .collection(id)
-                    .ok()
-                    .map(|state| state.write_frontier.clone());
-
-                if let (Some(old), Some(new)) = (old_upper, new_upper) {
-                    (old != new).then_some(ComputeControllerResponse::FrontierUpper { id, upper })
-                } else {
-                    // this is surprising, but we should already log something in `handle_frontier_upper`,
-                    // so no need to do so here
-                    None
-                }
+                self.handle_frontier_upper(id, upper.clone(), replica_id)

⬇️ Suggested change
-                self.handle_frontier_upper(id, upper.clone(), replica_id)
+                self.handle_frontier_upper(id, upper, replica_id)





In src/compute-client/src/controller/instance.rs:
>      fn handle_frontier_upper(
         &mut self,
         id: GlobalId,
         new_frontier: Antichain<T>,
         replica_id: ReplicaId,
-    ) {
+    ) -> Option<ComputeControllerResponse<T>> {
+        let old_upper = self
+            .compute
+            .collection(id)
+            .ok()
+            .map(|state| state.write_frontier.clone());

We are getting the collection in the next let-else, so if we move this block down we don't have to handle the absence case.



In src/controller/src/lib.rs:
> @@ -167,8 +168,23 @@ pub struct Controller<T = mz_repr::Timestamp> {
     /// Arguments for secrets readers.
     secrets_args: SecretsReaderCliArgs,
 
-    watch_sets: BTreeMap<GlobalId, Vec<Rc<(T, Box<dyn Any>)>>>,
+    /// A map associating a global ID to a vector of all the watch sets
+    /// that that ID is part of which have not yet been fulfilled for that object.
+    ///
+    /// See [`self.install_watch_se)`] for a description of watch sets.

⬇️ Suggested change
-    /// See [`self.install_watch_se)`] for a description of watch sets.
+    /// See [`self.install_watch_set`] for a description of watch sets.





In src/controller/src/lib.rs:
>  
+    /// A list of watch sets that were already fulfilled as soon as
+    /// they were installed, and thus that must be returned to the
+    /// client on the next call to [`self.process`].
+    ///
+    /// See [`self.install_watch_se)`] for a description of watch sets.

⬇️ Suggested change
-    /// See [`self.install_watch_se)`] for a description of watch sets.
+    /// See [`self.install_watch_set`] for a description of watch sets.





In src/controller/src/lib.rs:
>                      .entry(id)
                     .or_default()
                     .push(Rc::clone(&state));
             }
         }
     }
 
+    /// Process a pending response from the storage controller. If necessary,
+    /// return a higher-level response to our client.
+    ///
+    /// Precondition: The last readiness status generated must be `Readiness::Storage`.

Is this really a necessary precondition? I don't think anything goes wrong if it's not fulfilled, the storage controller just won't have any work to do.



In src/controller/src/lib.rs:
> +    /// Precondition: The last readiness status generated must be `Readiness::Storage`.
+    async fn process_storage_response(
+        &mut self,
+    ) -> Result<Option<ControllerResponse<T>>, anyhow::Error> {
+        let maybe_response = self.storage.process().await?;
+        Ok(maybe_response.and_then(
+            |mz_storage_client::controller::Response::FrontierUpdates(r)| {
+                self.handle_frontier_updates(&r)
+            },
+        ))
+    }
+
+    /// Process a pending response from the compute controller. If necessary,
+    /// return a higher-level response to our client.
+    ///
+    /// Precondition: The last readiness status generated must be `Readiness::Compute`.

Same here: I don't think this is a necessary precondition.

**Pull request merged**